Core/src/RequestHandler.php

```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * Copyright 2023 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Core;
+
+use Google\ApiCore\Serializer;
+
+/**
+ * @internal
+ * Responsible for forwarding the requests to their
+ * respective GAPIC methdos via the request wrapper.
+ */
+class RequestHandler
+{
+    use EmulatorTrait;
+    use RequestCallerTrait;
+
+    /**
+     * @var Serializer
+     */
+    private $serializer;
+
+    private $gapics;
+
+    /**
+     * @param array $config
+     */
+    public function __construct(
+        Serializer $serializer,
+        array $gapicClasses,
+        array $clientConfig = []
+    ) {
+        //@codeCoverageIgnoreStart
+        $this->serializer = $serializer;
```

Comment:
Why does this keep the serializer as a private property? It's never used, and is only passed into the `GapicRequestWrapper`. 

I see we use `getSerializer` in the pubsub PR (https://github.com/googleapis/google-cloud-php/pull/6872#pullrequestreview-1796570889), but this worries me - why is serialization being done in 3 places (`RequestHandler`, `GapicRequestWrapper`, and `PubSubClient`)? Ideally, it would only happen in one of these places.

Reply:
Essentially, we need this serializer wherever we encode/decode a protobuf message.
Why do we need the same instance everywhere? So that the `decodeMessage` call takes care of the transforms that are meant for that product.

In the [PubSub PR](https://github.com/googleapis/google-cloud-php/pull/6872/files#diff-aea1e0f86c268c5068649f432f2f0d3d5f67f78177c2b396e8780241d6bb6862R902-R913), you can see that we define the Serialzer to make modifications to `publish_time` and `expiration_time` fields for example.

1. In PubSubClient, the `getSerializer` is simply an abstraction to a block of code. It's not used as a getter anywhere else, we are defining the Serializer here, so it looked cleaner to me to abstract it in a prvt method. I can define this in the constructor if that seems better to you.

2. In the GapicRequestWrapper, we have the private property because we have calls to `encodeMessage` and we want to make sure, we use the correct Serializer. We don't have a getter here.

3. Why do we have the `getSerializer` in `RequestHandler`? Check out the resource classes like [Topic.php](https://github.com/googleapis/google-cloud-php/pull/6872/files#diff-d8bceda48a9492fc0e6bef6cf25ea6c7520bdbef78a0e4700034c4c8c1b237aaR556), [Subscription.php](https://github.com/googleapis/google-cloud-php/pull/6872/files#diff-c21815ead183d93472342f6e53240ce6272170ab7cb5e06545af509f939ed44cR415) etc. These have a lot of calls to `decodeMessage`/`encodeMessage`, we need a way to be able to access the same Serializer instance. I can pass around the serialzer instance, but that seems an anti pattern, given that I am already passing the request handler, which encompasses the serialzer.

If you recall, in a previous design we had the `PubSubSerialzer` which was meant to reduce these calls.

Reply:
I am not concerned that we use the same _instance_ of `PubSubSerializer` in 3 places, I am concerned that the actual serialization of messages is happening in 3 places. 

1. if `PubSubClient` only defines the serializer and does not use it, this seems okay to me
2. `GapicRequestWrapper` also seems okay to serialize here
3. This is what worries me for two reasons: 1, if these calls are being made all over resource classes like `Topic` and `Subscription`, we should be passing in a `Serializer` instead of bundling it with `RequestHandler`. 2, these serialization calls were previously abstracted in the "Connection" classes, but this design they're being handled directly by each resource class. This seems like an unnecessary (and even harmful) de-abstraction that will make the code more complicated.